extensions/src/data_bridge/index.ts

```diff
@@ -17,75 +19,107 @@ const plugin: JupyterFrontEndPlugin<void> = {
-                    data: data,
+                    action: "set-data",
+                    payload: data
                 },
                 "*"
             );
         };
 
+        /**
+         * Listen for messages from the host page, and update the data in the kernel
+         * @param event MessageEvent
+         */
         // @ts-ignore
-        window.requestDataFromHost = (variableName = "data") => {
-            window.parent.postMessage(
-                {
-                    type: "from-iframe-to-host",
-                    requestData: true,
-                    variableName,
-                },
-                "*"
-            );
-        };
-
-        window.addEventListener("message", async (event) => {
+        window.addEventListener("message", async (event: MessageEvent<IframeMessageSchema>) => {
             if (event.data.type === "from-host-to-iframe") {
-                let data = event.data.data;
-                let variableName = event.data.variableName || "data";
-                const dataJson = JSON.stringify(data);
-                const code = `
-  import json
-  ${variableName} = json.loads('${dataJson}')
-      `;
-                // Similar to https://jupyterlab.readthedocs.io/en/stable/api/classes/application.LabShell.html#currentWidget
-                // https://jupyterlite.readthedocs.io/en/latest/reference/api/ts/interfaces/jupyterlite_application.ISingleWidgetShell.html#currentwidget
-                const currentWidget = app.shell.currentWidget;
-
-                if (currentWidget instanceof NotebookPanel) {
-                    const notebookPanel = currentWidget;
-                    const kernel = notebookPanel.sessionContext.session?.kernel;
-                    if (kernel) {
-                        kernel.requestExecute({ code: code });
-                    } else {
-                        console.error("No active kernel found");
-                    }
-                } else {
-                    console.error("Current active widget is not a notebook");
-                }
+                // @ts-ignore
+                app.dataFromHost = JSON.stringify(event.data.payload);
+                //@ts-ignore
+                console.debug("Data from host received. app:", app.dataFromHost);
+                // Execute code in the kernel
+                const notebookPanel = notebookTracker.currentWidget;
+                await notebookPanel.sessionContext.ready;
+                const sessionContext = notebookPanel.sessionContext;
+                const kernel = sessionContext.session?.kernel;
+                // @ts-ignore
+                loadData(kernel, app.dataFromHost);
             }
         });
+
+        /**
+         * Load the data into the kernel by executing code
+         * @param kernel
+         * @param data
+         */
+        const loadData = (kernel: IKernelConnection, data: JSON) => {
+            const dataFromHostString = JSON.stringify(data);
```

Comment:
Same as above

Reply:
This is actually necessary to escape all the special characteres, so that this can be used directly in the Python code.